From patchwork Wed Dec  5 23:28:58 2018
Content-Type: text/plain; charset="utf-8"
MIME-Version: 1.0
Content-Transfer-Encoding: 7bit
X-Patchwork-Submitter: Jim Mattson <jmattson@google.com>
X-Patchwork-Id: 10715049
Return-Path: <kvm-owner@kernel.org>
Received: from mail.wl.linuxfoundation.org (pdx-wl-mail.web.codeaurora.org
 [172.30.200.125])
	by pdx-korg-patchwork-2.web.codeaurora.org (Postfix) with ESMTP id E17E313BB
	for <patchwork-kvm@patchwork.kernel.org>;
 Wed,  5 Dec 2018 23:29:13 +0000 (UTC)
Received: from mail.wl.linuxfoundation.org (localhost [127.0.0.1])
	by mail.wl.linuxfoundation.org (Postfix) with ESMTP id D373E2E0B8
	for <patchwork-kvm@patchwork.kernel.org>;
 Wed,  5 Dec 2018 23:29:13 +0000 (UTC)
Received: by mail.wl.linuxfoundation.org (Postfix, from userid 486)
	id C793A2E0CD; Wed,  5 Dec 2018 23:29:13 +0000 (UTC)
X-Spam-Checker-Version: SpamAssassin 3.3.1 (2010-03-16) on
	pdx-wl-mail.web.codeaurora.org
X-Spam-Level: 
X-Spam-Status: No, score=-15.5 required=2.0 tests=BAYES_00,DKIM_SIGNED,
	DKIM_VALID,DKIM_VALID_AU,MAILING_LIST_MULTI,RCVD_IN_DNSWL_HI,
	USER_IN_DEF_DKIM_WL autolearn=ham version=3.3.1
Received: from vger.kernel.org (vger.kernel.org [209.132.180.67])
	by mail.wl.linuxfoundation.org (Postfix) with ESMTP id 758952E0B8
	for <patchwork-kvm@patchwork.kernel.org>;
 Wed,  5 Dec 2018 23:29:13 +0000 (UTC)
Received: (majordomo@vger.kernel.org) by vger.kernel.org via listexpand
        id S1728052AbeLEX3M (ORCPT
        <rfc822;patchwork-kvm@patchwork.kernel.org>);
        Wed, 5 Dec 2018 18:29:12 -0500
Received: from mail-io1-f73.google.com ([209.85.166.73]:56584 "EHLO
        mail-io1-f73.google.com" rhost-flags-OK-OK-OK-OK) by vger.kernel.org
        with ESMTP id S1727592AbeLEX3M (ORCPT <rfc822;kvm@vger.kernel.org>);
        Wed, 5 Dec 2018 18:29:12 -0500
Received: by mail-io1-f73.google.com with SMTP id m1so21981650ioh.23
        for <kvm@vger.kernel.org>; Wed, 05 Dec 2018 15:29:11 -0800 (PST)
DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed;
        d=google.com; s=20161025;
        h=date:in-reply-to:message-id:mime-version:references:subject:from:to
         :cc;
        bh=YO6T/jppX4s0Vx9VP12nEwWUK8CfFZPsnJ4DzfreIWo=;
        b=E6DdPazb9PGE/n2hbqg+rnUd0N0A/OmSPA2WXks0ykGeKqFUZQAb5+QiaLD4y6an0A
         sqS1fimeJ3vA5v8yJ/j2GsdJ3crSuJDM6Ac90eJ8NK+tjq/xXvREzhFbrG6GhcOC7+F1
         zp6sVACXUIeN3gguFKHGcHXT0YfdFNfznR9769XDSsBey2d4xT26fGmN2oUdgB+6y1Ri
         uySjokBjlBEC4WnVZOtfmgYejRmbnm3kEKS/0SlOBJfKoWiwtq7eorJpy6K4h8TQ5GzJ
         nI1dkldRYLHv/3tScgLV8OBVqU4iaiGhtdcQIggZDWrEcM1yMxQkTqWdivtSEDCjjv+7
         RHfw==
X-Google-DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed;
        d=1e100.net; s=20161025;
        h=x-gm-message-state:date:in-reply-to:message-id:mime-version
         :references:subject:from:to:cc;
        bh=YO6T/jppX4s0Vx9VP12nEwWUK8CfFZPsnJ4DzfreIWo=;
        b=B4xX6Sgsxp91jVHqeQKIhqpHjl49aBFhKU4/h6xPSNxxp+2diZtBVVVjcQutHD29dK
         KoWCno2btzy4tC978i2LTWtltaDpRD1BFpdh4sFBVtYqngR+Vnke3ExYGlJa2OJzZ2Js
         eW3yrZ9n1XNrB5aRpfF75frDWdxnPnBaCVcFdqzQdKdF75+KXlQ9iwPgxdbsYr8Y61Pi
         FRyPJlI+wJG7w5WknWQCfLLx6484XaJQ2Tg1VKtdAZRyMr+w41x7zPILDvbBJ+mCrUky
         LSpJJHFZh/SxIFZ5kzmgBzjjdp3B/lSeanegddoXBtRtQGv342R96keH2cj77OW/1iMI
         b4Sg==
X-Gm-Message-State: AA+aEWZ/2u/xUIogYHCv7pq0W7YAZ+YzO+tCG8JbXl/crK2fXeLRiLIa
        2ILjHnmcHM6RkwZ0xXRS4E/Q/+4BWoUOkjXM6vWZxIcFyxVLaHb8x2aoSMT8Dp8Yv450Y56I3rY
        Re1H7/rwmfiXu0QQ6Q6LKccsUo4iUvZXy+AxgqtJr2iozTrK3RCsrj+EwTWoJnCc=
X-Google-Smtp-Source: 
 AFSGD/UVepC16o7Aty9ys/FtxmjcqhCtjsDnwERqL/IoMDHhuE+SY7+uVTxcgOqyeXyCf6qvPOLU5Zedv97vrQ==
X-Received: by 2002:a24:5755:: with SMTP id u82mr15392058ita.16.1544052551000;
 Wed, 05 Dec 2018 15:29:11 -0800 (PST)
Date: Wed,  5 Dec 2018 15:28:58 -0800
In-Reply-To: <20181205232902.1431-1-jmattson@google.com>
Message-Id: <20181205232902.1431-2-jmattson@google.com>
Mime-Version: 1.0
References: <20181205232902.1431-1-jmattson@google.com>
X-Mailer: git-send-email 2.20.0.rc1.387.gf8505762e3-goog
Subject: [PATCH v4 1/5] kvm: vmx: Set IA32_TSC_AUX for legacy mode guests
From: Jim Mattson <jmattson@google.com>
To: kvm@vger.kernel.org
Cc: Jim Mattson <jmattson@google.com>, Peter Shier <pshier@google.com>,
        Marc Orr <marcorr@google.com>,
        Liran Alon <liran.alon@oracle.com>
Content-Type: text/plain; charset="UTF-8"
Sender: kvm-owner@vger.kernel.org
Precedence: bulk
List-ID: <kvm.vger.kernel.org>
X-Mailing-List: kvm@vger.kernel.org
X-Virus-Scanned: ClamAV using ClamSMTP

RDTSCP is supported in legacy mode as well as long mode. The
IA32_TSC_AUX MSR should be set to the correct guest value before
entering any guest that supports RDTSCP.

Fixes: 4e47c7a6d714 ("KVM: VMX: Add instruction rdtscp support for guest")
Signed-off-by: Jim Mattson <jmattson@google.com>
Reviewed-by: Peter Shier <pshier@google.com>
Reviewed-by: Marc Orr <marcorr@google.com>
Reviewed-by: Liran Alon <liran.alon@oracle.com>
Reviewed-by: Sean Christopherson <sean.j.christopherson@intel.com>
---
 arch/x86/kvm/vmx.c | 6 +++---
 1 file changed, 3 insertions(+), 3 deletions(-)

diff --git a/arch/x86/kvm/vmx.c b/arch/x86/kvm/vmx.c
index c379d0bfdcba9..ad1753e8a85e5 100644
--- a/arch/x86/kvm/vmx.c
+++ b/arch/x86/kvm/vmx.c
@@ -3436,9 +3436,6 @@ static void setup_msrs(struct vcpu_vmx *vmx)
 		index = __find_msr_index(vmx, MSR_CSTAR);
 		if (index >= 0)
 			move_msr_up(vmx, index, save_nmsrs++);
-		index = __find_msr_index(vmx, MSR_TSC_AUX);
-		if (index >= 0 && guest_cpuid_has(&vmx->vcpu, X86_FEATURE_RDTSCP))
-			move_msr_up(vmx, index, save_nmsrs++);
 		/*
 		 * MSR_STAR is only needed on long mode guests, and only
 		 * if efer.sce is enabled.
@@ -3451,6 +3448,9 @@ static void setup_msrs(struct vcpu_vmx *vmx)
 	index = __find_msr_index(vmx, MSR_EFER);
 	if (index >= 0 && update_transition_efer(vmx, index))
 		move_msr_up(vmx, index, save_nmsrs++);
+	index = __find_msr_index(vmx, MSR_TSC_AUX);
+	if (index >= 0 && guest_cpuid_has(&vmx->vcpu, X86_FEATURE_RDTSCP))
+		move_msr_up(vmx, index, save_nmsrs++);
 
 	vmx->save_nmsrs = save_nmsrs;
 	vmx->guest_msrs_dirty = true;

From patchwork Wed Dec  5 23:28:59 2018
Content-Type: text/plain; charset="utf-8"
MIME-Version: 1.0
Content-Transfer-Encoding: 7bit
X-Patchwork-Submitter: Jim Mattson <jmattson@google.com>
X-Patchwork-Id: 10715051
Return-Path: <kvm-owner@kernel.org>
Received: from mail.wl.linuxfoundation.org (pdx-wl-mail.web.codeaurora.org
 [172.30.200.125])
	by pdx-korg-patchwork-2.web.codeaurora.org (Postfix) with ESMTP id CA94D1731
	for <patchwork-kvm@patchwork.kernel.org>;
 Wed,  5 Dec 2018 23:29:15 +0000 (UTC)
Received: from mail.wl.linuxfoundation.org (localhost [127.0.0.1])
	by mail.wl.linuxfoundation.org (Postfix) with ESMTP id BC6912E0B8
	for <patchwork-kvm@patchwork.kernel.org>;
 Wed,  5 Dec 2018 23:29:15 +0000 (UTC)
Received: by mail.wl.linuxfoundation.org (Postfix, from userid 486)
	id B0BE72E0DF; Wed,  5 Dec 2018 23:29:15 +0000 (UTC)
X-Spam-Checker-Version: SpamAssassin 3.3.1 (2010-03-16) on
	pdx-wl-mail.web.codeaurora.org
X-Spam-Level: 
X-Spam-Status: No, score=-15.5 required=2.0 tests=BAYES_00,DKIM_SIGNED,
	DKIM_VALID,DKIM_VALID_AU,MAILING_LIST_MULTI,RCVD_IN_DNSWL_HI,
	USER_IN_DEF_DKIM_WL autolearn=ham version=3.3.1
Received: from vger.kernel.org (vger.kernel.org [209.132.180.67])
	by mail.wl.linuxfoundation.org (Postfix) with ESMTP id 41CD32E0B8
	for <patchwork-kvm@patchwork.kernel.org>;
 Wed,  5 Dec 2018 23:29:15 +0000 (UTC)
Received: (majordomo@vger.kernel.org) by vger.kernel.org via listexpand
        id S1728209AbeLEX3O (ORCPT
        <rfc822;patchwork-kvm@patchwork.kernel.org>);
        Wed, 5 Dec 2018 18:29:14 -0500
Received: from mail-qt1-f201.google.com ([209.85.160.201]:45822 "EHLO
        mail-qt1-f201.google.com" rhost-flags-OK-OK-OK-OK) by vger.kernel.org
        with ESMTP id S1727794AbeLEX3O (ORCPT <rfc822;kvm@vger.kernel.org>);
        Wed, 5 Dec 2018 18:29:14 -0500
Received: by mail-qt1-f201.google.com with SMTP id w1so22054157qta.12
        for <kvm@vger.kernel.org>; Wed, 05 Dec 2018 15:29:13 -0800 (PST)
DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed;
        d=google.com; s=20161025;
        h=date:in-reply-to:message-id:mime-version:references:subject:from:to
         :cc;
        bh=L+LzLEys2DYGriE6mkCFFD365yssLfOW8pq2r8q5pxg=;
        b=oFcJGBIHOHgG4tXSrz63s5p88vVuoFA2hs0jQ8dP/2VYrvqTY2UN0s7IDyBE0rsmb+
         rci2QfYlR804BrZumxNGDCq8krOI1mRedqRFUxQx5hahmZPOAZFCQclY+kqUWSjpb/nB
         mi4Wv3Q6MMeNBKowffndI6IOxA49B6mMdJKlWPp6Ef/NEZJOLJ2yo1GnAYg3KkQ9Mu2I
         DU2qNkJRB+t2VHFxyzvZ+a67+3x4hJWCUt0s0KkAkqKUd2sspNN1ipM+MhNdO+q/YnAX
         4xWj+Tjdwu4urkoAi9G5PmFKVEzxv2uVHd1jykGNuHdn4vD4e2ufSzHGWg1gVD1K2Lf4
         3SyA==
X-Google-DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed;
        d=1e100.net; s=20161025;
        h=x-gm-message-state:date:in-reply-to:message-id:mime-version
         :references:subject:from:to:cc;
        bh=L+LzLEys2DYGriE6mkCFFD365yssLfOW8pq2r8q5pxg=;
        b=c0sTS0+rXgJee6ugUKYYCuQva/ECCwimETvB6UaV1WTnYXEISbQsToit6W+PBIf9uf
         WC1lM6/FBZiAe94MLTXhUo8/ZZmPdKXKLTsUE4cDdi5jify1Wysqnlobe6qeG0rBEu9d
         hiROuJcKk0F9AizWGIxeNDkHDyjfYRK3A0CBQbAuztpnc/4DBXf3ie1VWR7SrJFvqajk
         d/QOcq4lMroP5ULGvsNSfQwWGGQDh0l6osmI7WXbUDWGV0/q6tElxnRrbeenzC7LWWgb
         y+GD1176MLKM6ByIBbsN1cviZg/JwJ1oX0QfqUWLeaoooeWAkMUUaxGEAEfTLCGM8CMw
         87gA==
X-Gm-Message-State: AA+aEWYZixGo/wFxLQRlzywQhDhsUpnesNDbWBClPXpKkNq6w63z0Kwg
        sQBzH7Dg2Sk1Q2UsUw/CMaZq1FhaLHlutJgfcb3x9Ml65F3zPebVenzGs+5p41erWukOf2pQ5cQ
        Pg6dzLxEjTsoWBuDkDhc0vXVyyveO6n4ueUCWhWhz3pJbkuD6Z3q9aO0LeswzJHI=
X-Google-Smtp-Source: 
 AFSGD/WcW2PibWrsrY1tloT9Y5+SnilbcjYsIhLyq4TEJOiCk/8AV4i99jWM7/QemfV76qhjMJSmYz1VJqKkwg==
X-Received: by 2002:a0c:9ad6:: with SMTP id k22mr19557086qvf.7.1544052553103;
 Wed, 05 Dec 2018 15:29:13 -0800 (PST)
Date: Wed,  5 Dec 2018 15:28:59 -0800
In-Reply-To: <20181205232902.1431-1-jmattson@google.com>
Message-Id: <20181205232902.1431-3-jmattson@google.com>
Mime-Version: 1.0
References: <20181205232902.1431-1-jmattson@google.com>
X-Mailer: git-send-email 2.20.0.rc1.387.gf8505762e3-goog
Subject: [PATCH v4 2/5] kvm: vmx: Document the need for MSR_STAR in i386
 builds
From: Jim Mattson <jmattson@google.com>
To: kvm@vger.kernel.org
Cc: Jim Mattson <jmattson@google.com>, Peter Shier <pshier@google.com>
Content-Type: text/plain; charset="UTF-8"
Sender: kvm-owner@vger.kernel.org
Precedence: bulk
List-ID: <kvm.vger.kernel.org>
X-Mailing-List: kvm@vger.kernel.org
X-Virus-Scanned: ClamAV using ClamSMTP

Add a comment explaining why MSR_STAR must be included in
vmx_msr_index[] even for i386 builds.

The elided comment has not been relevant since move_msr_up() was
introduced in commit a75beee6e4f5d ("KVM: VMX: Avoid saving and
restoring msrs on lightweight vmexit").

Signed-off-by: Jim Mattson <jmattson@google.com>
Reviewed-by: Peter Shier <pshier@google.com>
---
 arch/x86/kvm/vmx.c | 7 +++++--
 1 file changed, 5 insertions(+), 2 deletions(-)

diff --git a/arch/x86/kvm/vmx.c b/arch/x86/kvm/vmx.c
index ad1753e8a85e5..b58c3952c5dfb 100644
--- a/arch/x86/kvm/vmx.c
+++ b/arch/x86/kvm/vmx.c
@@ -1390,8 +1390,11 @@ static u64 host_efer;
 static void ept_save_pdptrs(struct kvm_vcpu *vcpu);
 
 /*
- * Keep MSR_STAR at the end, as setup_msrs() will try to optimize it
- * away by decrementing the array size.
+ * Though SYSCALL is only supported in 64-bit mode on Intel CPUs, kvm
+ * will emulate SYSCALL in legacy mode if the vendor string in guest
+ * CPUID.0:{EBX,ECX,EDX} is "AuthenticAMD" or "AMDisbetter!" To
+ * support this emulation, IA32_STAR must always be included in
+ * vmx_msr_index[], even in i386 builds.
  */
 static const u32 vmx_msr_index[] = {
 #ifdef CONFIG_X86_64

From patchwork Wed Dec  5 23:29:00 2018
Content-Type: text/plain; charset="utf-8"
MIME-Version: 1.0
Content-Transfer-Encoding: 7bit
X-Patchwork-Submitter: Jim Mattson <jmattson@google.com>
X-Patchwork-Id: 10715053
Return-Path: <kvm-owner@kernel.org>
Received: from mail.wl.linuxfoundation.org (pdx-wl-mail.web.codeaurora.org
 [172.30.200.125])
	by pdx-korg-patchwork-2.web.codeaurora.org (Postfix) with ESMTP id 162D91731
	for <patchwork-kvm@patchwork.kernel.org>;
 Wed,  5 Dec 2018 23:29:18 +0000 (UTC)
Received: from mail.wl.linuxfoundation.org (localhost [127.0.0.1])
	by mail.wl.linuxfoundation.org (Postfix) with ESMTP id 07CB62E0B8
	for <patchwork-kvm@patchwork.kernel.org>;
 Wed,  5 Dec 2018 23:29:18 +0000 (UTC)
Received: by mail.wl.linuxfoundation.org (Postfix, from userid 486)
	id F05032E0DF; Wed,  5 Dec 2018 23:29:17 +0000 (UTC)
X-Spam-Checker-Version: SpamAssassin 3.3.1 (2010-03-16) on
	pdx-wl-mail.web.codeaurora.org
X-Spam-Level: 
X-Spam-Status: No, score=-15.5 required=2.0 tests=BAYES_00,DKIM_SIGNED,
	DKIM_VALID,DKIM_VALID_AU,MAILING_LIST_MULTI,RCVD_IN_DNSWL_HI,
	USER_IN_DEF_DKIM_WL autolearn=ham version=3.3.1
Received: from vger.kernel.org (vger.kernel.org [209.132.180.67])
	by mail.wl.linuxfoundation.org (Postfix) with ESMTP id 9F95A2E0B8
	for <patchwork-kvm@patchwork.kernel.org>;
 Wed,  5 Dec 2018 23:29:17 +0000 (UTC)
Received: (majordomo@vger.kernel.org) by vger.kernel.org via listexpand
        id S1728683AbeLEX3Q (ORCPT
        <rfc822;patchwork-kvm@patchwork.kernel.org>);
        Wed, 5 Dec 2018 18:29:16 -0500
Received: from mail-qk1-f202.google.com ([209.85.222.202]:33220 "EHLO
        mail-qk1-f202.google.com" rhost-flags-OK-OK-OK-OK) by vger.kernel.org
        with ESMTP id S1727794AbeLEX3Q (ORCPT <rfc822;kvm@vger.kernel.org>);
        Wed, 5 Dec 2018 18:29:16 -0500
Received: by mail-qk1-f202.google.com with SMTP id 80so21756356qkd.0
        for <kvm@vger.kernel.org>; Wed, 05 Dec 2018 15:29:15 -0800 (PST)
DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed;
        d=google.com; s=20161025;
        h=date:in-reply-to:message-id:mime-version:references:subject:from:to
         :cc;
        bh=r5aDJDBOexkPknRaeDrpjroigSkuctKmlwLBNy1If9k=;
        b=I/tGu8MHrJjQOWXxtEXXhQuz/+UdU6qlQEmfbC0hDbYC3I17lm8kXDUs8bEVXV//je
         0PJKO6nw+JsJ6RafGJUMU92rKX5IEhy4d6u2fAaLf/MkqWfr+9jOHsph7c0zxMU0i7vs
         aYUatc2HvO3cATvsODHv9faYktWi4u1+urZ6DElgiPzKipeNxi9voLmeOBudAABSlIx0
         5hGTHksgcC9e5QW+C0LSRdGSMQW+ann7P+cpoXUnEXwD39ykjMCVgxNxOtx8p6qOEeDa
         u/w0QkX4IG5UCSo5J3HqjXtvjDq1zo5KwAoYakVq3hIQFJEiJ46wm9So7BnsTU4qZHUf
         3O/g==
X-Google-DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed;
        d=1e100.net; s=20161025;
        h=x-gm-message-state:date:in-reply-to:message-id:mime-version
         :references:subject:from:to:cc;
        bh=r5aDJDBOexkPknRaeDrpjroigSkuctKmlwLBNy1If9k=;
        b=V2/C+UuARR8z7snziDlEzIBopNhPwnYSVksrygDWdA9jvRn6+jwH2D0Sfk6znifyuF
         ETSfatiZqYcx01e244Lc7gW2KEUZvY5Sp6zgKNSV6QDr6cIzJpdhTAHrg4tSqPeNoppV
         Wh3CwkiP+HifMFDtlxj4FVjS7yFjRdTHu6UN6XzydcCKELUtSBBOgDveKmpUleCUVmV1
         jOlPGNrixSJqKuePBnCiM7Qv+mOU6K1H6+FpOtCF9JrbNjHpRKxWos2HTkVfXdlGPuHk
         q8vN+ErPSftDBFRSZzoty2WP7NiZwNssUyUcCUILo5xBsdCBMFc+IFDY/bwfFcMGM+ca
         FHWw==
X-Gm-Message-State: AA+aEWbJnhw+Y8v1CnZVuNihHT1SL3FZdI8uQWy4LFwq+Djs0yi9XEOP
        rG0mSHHbEI2HGqQj1i7vOcLSaT13OJ8lGKqfCRDKHhTFVChcut9PNp7vgBPYBPxkNVShrZRyG13
        U06jcaITCIEFi00ziVqmisWesP2xipgj88cnrXISdQ4i07zjpPXewckHAJLLnXa0=
X-Google-Smtp-Source: 
 AFSGD/Uw2h3Fm9VmR0FwrDtZJlNN4qf9MPCWgvdK+BJ2blrf+KHETcTpNxdLSEldS7rMK288dasDSdSY4zRQCg==
X-Received: by 2002:a37:4cd0:: with SMTP id
 z199mr18765521qka.44.1544052555201;
 Wed, 05 Dec 2018 15:29:15 -0800 (PST)
Date: Wed,  5 Dec 2018 15:29:00 -0800
In-Reply-To: <20181205232902.1431-1-jmattson@google.com>
Message-Id: <20181205232902.1431-4-jmattson@google.com>
Mime-Version: 1.0
References: <20181205232902.1431-1-jmattson@google.com>
X-Mailer: git-send-email 2.20.0.rc1.387.gf8505762e3-goog
Subject: [PATCH v4 3/5] kvm: vmx: Don't set hardware IA32_CSTAR MSR on
 VM-entry
From: Jim Mattson <jmattson@google.com>
To: kvm@vger.kernel.org
Cc: Jim Mattson <jmattson@google.com>, Peter Shier <pshier@google.com>
Content-Type: text/plain; charset="UTF-8"
Sender: kvm-owner@vger.kernel.org
Precedence: bulk
List-ID: <kvm.vger.kernel.org>
X-Mailing-List: kvm@vger.kernel.org
X-Virus-Scanned: ClamAV using ClamSMTP

SYSCALL raises #UD in compatibility mode on Intel CPUs, so it's
pointless to load the guest's IA32_CSTAR value into the hardware MSR.

IA32_CSTAR still provides 48 bits of storage on Intel CPUs that have
CPUID.80000001:EDX.LM[bit 29] set, so we cannot remove it from the
vmx_msr_index[] array.

Signed-off-by: Jim Mattson <jmattson@google.com>
Reviewed-by: Peter Shier <pshier@google.com>
---
 arch/x86/kvm/vmx.c | 3 ---
 1 file changed, 3 deletions(-)

diff --git a/arch/x86/kvm/vmx.c b/arch/x86/kvm/vmx.c
index b58c3952c5dfb..12430486e911e 100644
--- a/arch/x86/kvm/vmx.c
+++ b/arch/x86/kvm/vmx.c
@@ -3434,9 +3434,6 @@ static void setup_msrs(struct vcpu_vmx *vmx)
 		if (index >= 0)
 			move_msr_up(vmx, index, save_nmsrs++);
 		index = __find_msr_index(vmx, MSR_LSTAR);
-		if (index >= 0)
-			move_msr_up(vmx, index, save_nmsrs++);
-		index = __find_msr_index(vmx, MSR_CSTAR);
 		if (index >= 0)
 			move_msr_up(vmx, index, save_nmsrs++);
 		/*

From patchwork Wed Dec  5 23:29:01 2018
Content-Type: text/plain; charset="utf-8"
MIME-Version: 1.0
Content-Transfer-Encoding: 7bit
X-Patchwork-Submitter: Jim Mattson <jmattson@google.com>
X-Patchwork-Id: 10715055
Return-Path: <kvm-owner@kernel.org>
Received: from mail.wl.linuxfoundation.org (pdx-wl-mail.web.codeaurora.org
 [172.30.200.125])
	by pdx-korg-patchwork-2.web.codeaurora.org (Postfix) with ESMTP id 2645913BB
	for <patchwork-kvm@patchwork.kernel.org>;
 Wed,  5 Dec 2018 23:29:22 +0000 (UTC)
Received: from mail.wl.linuxfoundation.org (localhost [127.0.0.1])
	by mail.wl.linuxfoundation.org (Postfix) with ESMTP id 17F4A2E0B8
	for <patchwork-kvm@patchwork.kernel.org>;
 Wed,  5 Dec 2018 23:29:22 +0000 (UTC)
Received: by mail.wl.linuxfoundation.org (Postfix, from userid 486)
	id 0C4292E0F5; Wed,  5 Dec 2018 23:29:22 +0000 (UTC)
X-Spam-Checker-Version: SpamAssassin 3.3.1 (2010-03-16) on
	pdx-wl-mail.web.codeaurora.org
X-Spam-Level: 
X-Spam-Status: No, score=-15.5 required=2.0 tests=BAYES_00,DKIM_SIGNED,
	DKIM_VALID,DKIM_VALID_AU,MAILING_LIST_MULTI,RCVD_IN_DNSWL_HI,
	USER_IN_DEF_DKIM_WL autolearn=ham version=3.3.1
Received: from vger.kernel.org (vger.kernel.org [209.132.180.67])
	by mail.wl.linuxfoundation.org (Postfix) with ESMTP id 141112E0BE
	for <patchwork-kvm@patchwork.kernel.org>;
 Wed,  5 Dec 2018 23:29:20 +0000 (UTC)
Received: (majordomo@vger.kernel.org) by vger.kernel.org via listexpand
        id S1728744AbeLEX3S (ORCPT
        <rfc822;patchwork-kvm@patchwork.kernel.org>);
        Wed, 5 Dec 2018 18:29:18 -0500
Received: from mail-it1-f202.google.com ([209.85.166.202]:38984 "EHLO
        mail-it1-f202.google.com" rhost-flags-OK-OK-OK-OK) by vger.kernel.org
        with ESMTP id S1728411AbeLEX3S (ORCPT <rfc822;kvm@vger.kernel.org>);
        Wed, 5 Dec 2018 18:29:18 -0500
Received: by mail-it1-f202.google.com with SMTP id k133so18292260ite.4
        for <kvm@vger.kernel.org>; Wed, 05 Dec 2018 15:29:17 -0800 (PST)
DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed;
        d=google.com; s=20161025;
        h=date:in-reply-to:message-id:mime-version:references:subject:from:to
         :cc;
        bh=+fjSOfpO/GDAmXM+nGFkTRaHaV5WH4ZMDHZldXRn8sk=;
        b=WrHzm5VfnCJvfS0xr5ZYjy1t8QJn0EYV2O/DBcFGL6LEd88vNA32aL+wBP9wU9etDH
         L/gMjm6YhCG8G4ugvqz6mNxBDQ074k5Tvn5ZyXTX/kgvDQxK+oYdvOKX/kX2gJR6whDK
         7g82z9BGCbdBNemp3Qs7BDYdle37sXWYUqhOiMlByz/J/+HlK/YA8uRsTVAWTCbSzzt6
         h9EOoo3FCMiWwxnodRtN6ao5lDfPX1A4FC4Kk8mKQ0I6XF4xlgSFAYkUgJHNptMW6kJT
         kxdCU1dzX8wEC65teJXMEFyoxoFgmIyO+hbDfgvPqqqGNEXyDzqpuDgq9LiSu4L3TVOO
         UsLQ==
X-Google-DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed;
        d=1e100.net; s=20161025;
        h=x-gm-message-state:date:in-reply-to:message-id:mime-version
         :references:subject:from:to:cc;
        bh=+fjSOfpO/GDAmXM+nGFkTRaHaV5WH4ZMDHZldXRn8sk=;
        b=GIWYEnZDasfWyivB3a/xloXzNc+HkjqMwTev7nrr1s+PmSh10WJbOJGjMWtUwfyGOl
         6SCQOg+pogrcBjUY8oruzAIauDdVZqFUACPzpBl48F4HpAcFdB7KqFd+VLPT4vqvAvOp
         OF3iBwZ9stOI6eJIDKXAzXcbionttOXhAburwKYV7x582fRkL29a1Rv717QLPU9r7Z0Z
         7UmBJ2lSatpjb2Uw0fJO/RrnVnBO+C26juYm/Lj03K+700FcLzmKF4d0pHKnF7InMgQv
         0QsHT+mh/tW1QQvLTljDhzG1x53xyTJnB3UdP07H8Cjo52UBokGLr/BuHlfyxotlKm7T
         q/yA==
X-Gm-Message-State: AA+aEWYgBKFORhrmfU66CfNAZOoICS4qajQ0/W5176xBll3JYEnvpyI6
        78ypUU/EgHNwJj3ci/T0/R4AVLMNn7ZkUFigcycX8LxP3ny97eYlIXE+bP23lbvuY2ncFk/ByfN
        B7w6489DN1bg8aWfBaqEa65C6XDN5956Ffs1erg2qfwwAVtwVbM90/rVF0U8p/kA=
X-Google-Smtp-Source: 
 AFSGD/XbuLVkA34A9ScQWF+OHW1bu7WPuA7p+v5vBSu6mc902cO4ZB/7Fip+b+zxn2HazWr8B72mGM5qS0YhIA==
X-Received: by 2002:a24:5f49:: with SMTP id r70mr15413147itb.22.1544052557155;
 Wed, 05 Dec 2018 15:29:17 -0800 (PST)
Date: Wed,  5 Dec 2018 15:29:01 -0800
In-Reply-To: <20181205232902.1431-1-jmattson@google.com>
Message-Id: <20181205232902.1431-5-jmattson@google.com>
Mime-Version: 1.0
References: <20181205232902.1431-1-jmattson@google.com>
X-Mailer: git-send-email 2.20.0.rc1.387.gf8505762e3-goog
Subject: [PATCH v4 4/5] kvm: vmx: Skip all SYSCALL MSRs in setup_msrs() when
 !EFER.SCE
From: Jim Mattson <jmattson@google.com>
To: kvm@vger.kernel.org
Cc: Jim Mattson <jmattson@google.com>, Peter Shier <pshier@google.com>,
        Marc Orr <marcorr@google.com>
Content-Type: text/plain; charset="UTF-8"
Sender: kvm-owner@vger.kernel.org
Precedence: bulk
List-ID: <kvm.vger.kernel.org>
X-Mailing-List: kvm@vger.kernel.org
X-Virus-Scanned: ClamAV using ClamSMTP

Like IA32_STAR, IA32_LSTAR and IA32_FMASK only need to contain guest
values on VM-entry when the guest is in long mode and EFER.SCE is set.

Signed-off-by: Jim Mattson <jmattson@google.com>
Reviewed-by: Peter Shier <pshier@google.com>
Reviewed-by: Marc Orr <marcorr@google.com>
---
 arch/x86/kvm/vmx.c | 16 ++++++++--------
 1 file changed, 8 insertions(+), 8 deletions(-)

diff --git a/arch/x86/kvm/vmx.c b/arch/x86/kvm/vmx.c
index 12430486e911e..7ae3842f659e2 100644
--- a/arch/x86/kvm/vmx.c
+++ b/arch/x86/kvm/vmx.c
@@ -3429,19 +3429,19 @@ static void setup_msrs(struct vcpu_vmx *vmx)
 
 	save_nmsrs = 0;
 #ifdef CONFIG_X86_64
-	if (is_long_mode(&vmx->vcpu)) {
-		index = __find_msr_index(vmx, MSR_SYSCALL_MASK);
+	/*
+	 * The SYSCALL MSRs are only needed on long mode guests, and only
+	 * when EFER.SCE is set.
+	 */
+	if (is_long_mode(&vmx->vcpu) && (vmx->vcpu.arch.efer & EFER_SCE)) {
+		index = __find_msr_index(vmx, MSR_STAR);
 		if (index >= 0)
 			move_msr_up(vmx, index, save_nmsrs++);
 		index = __find_msr_index(vmx, MSR_LSTAR);
 		if (index >= 0)
 			move_msr_up(vmx, index, save_nmsrs++);
-		/*
-		 * MSR_STAR is only needed on long mode guests, and only
-		 * if efer.sce is enabled.
-		 */
-		index = __find_msr_index(vmx, MSR_STAR);
-		if ((index >= 0) && (vmx->vcpu.arch.efer & EFER_SCE))
+		index = __find_msr_index(vmx, MSR_SYSCALL_MASK);
+		if (index >= 0)
 			move_msr_up(vmx, index, save_nmsrs++);
 	}
 #endif

From patchwork Wed Dec  5 23:29:02 2018
Content-Type: text/plain; charset="utf-8"
MIME-Version: 1.0
Content-Transfer-Encoding: 7bit
X-Patchwork-Submitter: Jim Mattson <jmattson@google.com>
X-Patchwork-Id: 10715057
Return-Path: <kvm-owner@kernel.org>
Received: from mail.wl.linuxfoundation.org (pdx-wl-mail.web.codeaurora.org
 [172.30.200.125])
	by pdx-korg-patchwork-2.web.codeaurora.org (Postfix) with ESMTP id 8902718A7
	for <patchwork-kvm@patchwork.kernel.org>;
 Wed,  5 Dec 2018 23:29:22 +0000 (UTC)
Received: from mail.wl.linuxfoundation.org (localhost [127.0.0.1])
	by mail.wl.linuxfoundation.org (Postfix) with ESMTP id 7899B2E0B8
	for <patchwork-kvm@patchwork.kernel.org>;
 Wed,  5 Dec 2018 23:29:22 +0000 (UTC)
Received: by mail.wl.linuxfoundation.org (Postfix, from userid 486)
	id 68D5F2E0BE; Wed,  5 Dec 2018 23:29:22 +0000 (UTC)
X-Spam-Checker-Version: SpamAssassin 3.3.1 (2010-03-16) on
	pdx-wl-mail.web.codeaurora.org
X-Spam-Level: 
X-Spam-Status: No, score=-15.5 required=2.0 tests=BAYES_00,DKIM_SIGNED,
	DKIM_VALID,DKIM_VALID_AU,MAILING_LIST_MULTI,RCVD_IN_DNSWL_HI,
	USER_IN_DEF_DKIM_WL autolearn=ham version=3.3.1
Received: from vger.kernel.org (vger.kernel.org [209.132.180.67])
	by mail.wl.linuxfoundation.org (Postfix) with ESMTP id 96A6C2E0F0
	for <patchwork-kvm@patchwork.kernel.org>;
 Wed,  5 Dec 2018 23:29:21 +0000 (UTC)
Received: (majordomo@vger.kernel.org) by vger.kernel.org via listexpand
        id S1728695AbeLEX3U (ORCPT
        <rfc822;patchwork-kvm@patchwork.kernel.org>);
        Wed, 5 Dec 2018 18:29:20 -0500
Received: from mail-oi1-f202.google.com ([209.85.167.202]:42810 "EHLO
        mail-oi1-f202.google.com" rhost-flags-OK-OK-OK-OK) by vger.kernel.org
        with ESMTP id S1728745AbeLEX3T (ORCPT <rfc822;kvm@vger.kernel.org>);
        Wed, 5 Dec 2018 18:29:19 -0500
Received: by mail-oi1-f202.google.com with SMTP id h85so13353987oib.9
        for <kvm@vger.kernel.org>; Wed, 05 Dec 2018 15:29:19 -0800 (PST)
DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed;
        d=google.com; s=20161025;
        h=date:in-reply-to:message-id:mime-version:references:subject:from:to
         :cc;
        bh=+1MlVB21bFxZDGSl/11a26XHdIGnetfG7HZxmxpHiP4=;
        b=Edi+FrfS650D8SOSd3FB1LYKU8hd6/+EsI6rnfSYFT9nqBWBuxzTt8V8F19WsWHv1R
         /047RbvGtzC+3E44oAWNkeEgYVZT3auD5SjzbP17s57rNHXuEBC+eendHh6GKjlwYX7D
         CyAO9lLjTVA6zD0vu0Oghm77Pqa6nH98nwBYCNtxBViBkN5Vc50qbOvd5rJFRy0iUjng
         tCJ+66x3gL6LmJyrpIDA1KV5N8OgBnm04Qk7OltSUU9jsh90AmpdhN/S4GDqxH9U1Jfm
         2Fbwxs4nGoaY3Q/hG+howNkUfx5QJ2mZoJ6Phbj9v57QaYuha5QAQURptqK5bv69uczU
         kvbw==
X-Google-DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed;
        d=1e100.net; s=20161025;
        h=x-gm-message-state:date:in-reply-to:message-id:mime-version
         :references:subject:from:to:cc;
        bh=+1MlVB21bFxZDGSl/11a26XHdIGnetfG7HZxmxpHiP4=;
        b=BoEsEd3Zxh+byebTFtFiBN1APEthNVU7z/6CPVTCbzkjEbPg+D8J8wLIaZ4xieX4RH
         PwlF2EDxKGm4xS9LcyBl+ozTaNdvOEzNLyr0XQSJ4TGwHOLM76Rt2pWrTMAdt+CXOHBr
         OvO9eh1zrpDMcbQix4MNzgSAW8olP9GvOIioi/kT57gr5e5vryoU2BSuj2dAzgC4miZR
         wbTquuN72VvBBhIZ0APf4zebryXq/KnXPQTdD0sGRSVPcIrT7xnZth+m5bIQKSMoKX64
         1bEsZVQqgPiKIZt1HMiA6ucBxkggc3fd2XLbB7CPojd4XVMrjYOvxFGvXG3G1TNx3fkM
         +xpw==
X-Gm-Message-State: AA+aEWbaKS5fUIm73nNTJfGH/fWdRTaqpb9uaX2EfNlGPYvOSBwjBiDV
        YFiA/YeKcq+0eK0diXh5gMy+Lsat6crhda4ntaEy9m7Y0touoDsC4yXFcSIkOEhjDF2aUoiT8EA
        8jDIfeeJOpq2MhzJ/jWP/6dq02FJq1Ow46Nv3hWt+2Ekg5vxxDG1AVhGiaeozYnE=
X-Google-Smtp-Source: 
 AFSGD/UHa/VcuWoPD+6rN5rsxeJXbV3G89VDfDl/WDDuqHuPheRyfcSBnOFDdx2Nq/5OMqdD6x2fMeJ0Y0fDWw==
X-Received: by 2002:aca:a949:: with SMTP id s70mr20818941oie.1.1544052559170;
 Wed, 05 Dec 2018 15:29:19 -0800 (PST)
Date: Wed,  5 Dec 2018 15:29:02 -0800
In-Reply-To: <20181205232902.1431-1-jmattson@google.com>
Message-Id: <20181205232902.1431-6-jmattson@google.com>
Mime-Version: 1.0
References: <20181205232902.1431-1-jmattson@google.com>
X-Mailer: git-send-email 2.20.0.rc1.387.gf8505762e3-goog
Subject: [PATCH v4 5/5] kvm: vmx: Stop wasting a page for guest_msrs
From: Jim Mattson <jmattson@google.com>
To: kvm@vger.kernel.org
Cc: Jim Mattson <jmattson@google.com>, Marc Orr <marcorr@google.com>,
        Peter Shier <pshier@google.com>
Content-Type: text/plain; charset="UTF-8"
Sender: kvm-owner@vger.kernel.org
Precedence: bulk
List-ID: <kvm.vger.kernel.org>
X-Mailing-List: kvm@vger.kernel.org
X-Virus-Scanned: ClamAV using ClamSMTP

We will never need more guest_msrs than there are indices in
vmx_msr_index.

Signed-off-by: Jim Mattson <jmattson@google.com>
Reviewed-by: Marc Orr <marcorr@google.com>
Reviewed-by: Peter Shier <pshier@google.com>
---
 arch/x86/kvm/vmx.c | 42 ++++++++++++++++--------------------------
 1 file changed, 16 insertions(+), 26 deletions(-)

diff --git a/arch/x86/kvm/vmx.c b/arch/x86/kvm/vmx.c
index 7ae3842f659e2..cf3c9eb4b4dcb 100644
--- a/arch/x86/kvm/vmx.c
+++ b/arch/x86/kvm/vmx.c
@@ -974,6 +974,20 @@ struct vmx_msrs {
 	struct vmx_msr_entry	val[NR_AUTOLOAD_MSRS];
 };
 
+/*
+ * Though SYSCALL is only supported in 64-bit mode on Intel CPUs, kvm
+ * will emulate SYSCALL in legacy mode if the vendor string in guest
+ * CPUID.0:{EBX,ECX,EDX} is "AuthenticAMD" or "AMDisbetter!" To
+ * support this emulation, IA32_STAR must always be included in
+ * vmx_msr_index[], even in i386 builds.
+ */
+static const u32 vmx_msr_index[] = {
+#ifdef CONFIG_X86_64
+	MSR_SYSCALL_MASK, MSR_LSTAR, MSR_CSTAR,
+#endif
+	MSR_EFER, MSR_TSC_AUX, MSR_STAR,
+};
+
 struct vcpu_vmx {
 	struct kvm_vcpu       vcpu;
 	unsigned long         host_rsp;
@@ -982,7 +996,7 @@ struct vcpu_vmx {
 	u32                   exit_intr_info;
 	u32                   idt_vectoring_info;
 	ulong                 rflags;
-	struct shared_msr_entry *guest_msrs;
+	struct shared_msr_entry guest_msrs[ARRAY_SIZE(vmx_msr_index)];
 	int                   nmsrs;
 	int                   save_nmsrs;
 	bool                  guest_msrs_dirty;
@@ -1389,20 +1403,6 @@ static u64 host_efer;
 
 static void ept_save_pdptrs(struct kvm_vcpu *vcpu);
 
-/*
- * Though SYSCALL is only supported in 64-bit mode on Intel CPUs, kvm
- * will emulate SYSCALL in legacy mode if the vendor string in guest
- * CPUID.0:{EBX,ECX,EDX} is "AuthenticAMD" or "AMDisbetter!" To
- * support this emulation, IA32_STAR must always be included in
- * vmx_msr_index[], even in i386 builds.
- */
-static const u32 vmx_msr_index[] = {
-#ifdef CONFIG_X86_64
-	MSR_SYSCALL_MASK, MSR_LSTAR, MSR_CSTAR,
-#endif
-	MSR_EFER, MSR_TSC_AUX, MSR_STAR,
-};
-
 DEFINE_STATIC_KEY_FALSE(enable_evmcs);
 
 #define current_evmcs ((struct hv_enlightened_vmcs *)this_cpu_read(current_vmcs))
@@ -11524,7 +11524,6 @@ static void vmx_free_vcpu(struct kvm_vcpu *vcpu)
 	leave_guest_mode(vcpu);
 	vmx_free_vcpu_nested(vcpu);
 	free_loaded_vmcs(vmx->loaded_vmcs);
-	kfree(vmx->guest_msrs);
 	kvm_vcpu_uninit(vcpu);
 	kmem_cache_free(kvm_vcpu_cache, vmx);
 }
@@ -11559,16 +11558,9 @@ static struct kvm_vcpu *vmx_create_vcpu(struct kvm *kvm, unsigned int id)
 			goto uninit_vcpu;
 	}
 
-	vmx->guest_msrs = kmalloc(PAGE_SIZE, GFP_KERNEL);
-	BUILD_BUG_ON(ARRAY_SIZE(vmx_msr_index) * sizeof(vmx->guest_msrs[0])
-		     > PAGE_SIZE);
-
-	if (!vmx->guest_msrs)
-		goto free_pml;
-
 	err = alloc_loaded_vmcs(&vmx->vmcs01);
 	if (err < 0)
-		goto free_msrs;
+		goto free_pml;
 
 	msr_bitmap = vmx->vmcs01.msr_bitmap;
 	vmx_disable_intercept_for_msr(msr_bitmap, MSR_FS_BASE, MSR_TYPE_RW);
@@ -11618,8 +11610,6 @@ static struct kvm_vcpu *vmx_create_vcpu(struct kvm *kvm, unsigned int id)
 
 free_vmcs:
 	free_loaded_vmcs(vmx->loaded_vmcs);
-free_msrs:
-	kfree(vmx->guest_msrs);
 free_pml:
 	vmx_destroy_pml_buffer(vmx);
 uninit_vcpu:
